lib/packwerk/parse_run.rb

```diff
@@ -49,23 +49,30 @@ def update_deprecations
 
     def check
       reference_lister = CheckingDeprecatedReferences.new(@configuration.root_path)
-      offenses = run(reference_lister)
+      new_offenses = filtered_offenses(reference_lister)
 
-      message = @offenses_formatter.show_offenses(offenses)
-      Result.new(message: message, status: offenses.empty?)
+      message = @offenses_formatter.show_offenses(new_offenses)
+      Result.new(message: message, status: new_offenses.empty?)
     end
 
     private
 
-    def run(reference_lister)
+    def filtered_offenses(reference_lister, show_errors: true)
+      find_offenses(reference_lister, show_errors: show_errors).select do |offense|
+        unlisted?(offense, reference_lister)
+      end
+    end
+
+    def find_offenses(reference_lister, show_errors:)
       @progress_formatter.started(@files)
 
-      run_context = Packwerk::RunContext.from_configuration(@configuration, reference_lister: reference_lister)
+      run_context = Packwerk::RunContext.from_configuration(@configuration)
       all_offenses = T.let([], T.untyped)
       execution_time = Benchmark.realtime do
         @files.each do |path|
           run_context.process_file(file: path).tap do |offenses|
-            mark_progress(offenses: offenses, progress_formatter: @progress_formatter)
+            includes_unlisted = show_errors && offenses.any? { |offense| unlisted?(offense, reference_lister) }
```

Comment:
This is still iterating over the same list twice. Can we filter here and remove the `select` from `filtered_offenses?`. Something like:

```ruby
offenses = offenses.select { |offense|  unlisted?(offense, reference_lister) }
includes_unlisted = show_errors && !offenses.empty?
update_progress(failed: includes_unlisted)
all_offenses.concat(offenses)
```

I don't think `find_offenses` is called in a context where we we want the ones that are listed.

Reply:
Yeah you are right in the context of this PR, did you see my [comments above](https://github.com/Shopify/packwerk/pull/124#discussion_r642648570)? Basically I can pull the extra iteration now but it'll be coming back pretty soon once parallel is in.